microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java

```diff
@@ -0,0 +1,648 @@
+    private static final AtomicInteger WORKER_COUNTER = new AtomicInteger();
+
+    // ---------------------------------------------------------------------
+    private final ZContext ctx;
+    private final Socket internalRouterSocket;
+    private final Socket internalServiceSocket;
+    private final List<Socket> routerSockets = new ArrayList<>(); // Sockets for clients & public external workers
+    private final AtomicBoolean run = new AtomicBoolean(true);
+    private final SortedSet<RbacRole<?>> rbacRoles;
+    private final Map<String, Service> services = new HashMap<>(); // known services Map<'service name', Service>
+    private final Map<String, Worker> workers = new HashMap<>(); // known workers Map<addressHex, Worker>
+    private final Map<String, Client> clients = new HashMap<>();
+
+    private final Deque<Worker> waiting = new ArrayDeque<>(); // idle workers
+    private long heartbeatAt = System.currentTimeMillis() + HEARTBEAT_INTERVAL; // When to send HEARTBEAT
+
+    /**
+     * Initialize broker state.
+     * @param ioThreads number of threads dedicated to network IO (recommendation 1 thread per 1 GBit/s)
+     * @param nInternalThreads number of threads per internal service
+     * @param rbacRoles RBAC-based roles (used for IO prioritisation and service access control
+     */
+    public MajordomoBroker(final int ioThreads, final int nInternalThreads, final RbacRole<?>... rbacRoles) {
+        this.setName(MajordomoBroker.class.getSimpleName() + "#" + BROKER_COUNTER.getAndIncrement());
+
+        ctx = new ZContext(ioThreads);
+
+        // initialise RBAC role-based priority queues
+        this.rbacRoles = Collections.unmodifiableSortedSet(new TreeSet<>(Set.of(rbacRoles)));
+
+        // generate and register internal default inproc socket
+        this.internalRouterSocket = bind("inproc://broker"); // NOPMD
+        //this.internalRouterSocket.setRouterMandatory(true);
+        this.internalServiceSocket = bind("inproc://intService"); // NOPMD
+        //this.internalServiceSocket.setRouterMandatory(true);
+
+        registerDefaultServices(rbacRoles); // NOPMD
+    }
+
+    public void addInternalService(final MajordomoWorker worker, final int nServiceThreads) {
+        assert worker != null : "worker must not be null";
+        final Service oldWorker = services.put(worker.getServiceName(), new Service(worker.getServiceName(), worker.getServiceName().getBytes(StandardCharsets.UTF_8), worker, nServiceThreads));
+        if (oldWorker != null) {
+            LOGGER.atWarn().addArgument(worker.getServiceName()).log("overwriting existing internal service definition for '{}'");
+        }
+    }
+
+    /**
+     * Bind broker to endpoint, can call this multiple times. We use a single
+     * socket for both clients and workers.
+     */
+    public ZMQ.Socket bind(String endpoint) {
+        final ZMQ.Socket routerSocket = ctx.createSocket(SocketType.ROUTER);
```

Comment:
Codacy found an issue: [Unnecessary use of fully qualified name 'ZMQ.Socket' due to existing static import 'org.zeromq.ZMQ.Socket'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)